0001-typedb/03-admin/01-configuration.md

```diff
@@ -0,0 +1,248 @@
+    * `data`: cache for often-used data.
+    * `index`: cache for data indexes.
+
+<div class="note">
+[Important]
+For production use, it is recommended that the `server.data` is set to a path outside of the `$TYPEDB_HOME` (directory 
+with TypeDB server files). This helps to make the process of upgrading TypeDB easier.
+</div>
+
+If the index cache is too small relative to your dataset, you may find severely degraded performance. We recommend 
+allocating at least **2%** of your database size equivalent to the index cache. For example, with **100 GB** of 
+on-disk data in a database, allocate at least **2 GB** of index cache. Allocating more can improve performance.
+
+Additionally, we recommend the sum of data and index caches equal to about **20%** of the total memory of the server.
+
+### Logging configuration
+
+The log section of the configuration file contains the logging behavior options of TypeDB.
+
+There are three subsections:
+
+* `output`: Define destinations to write logs to. Allowed types are `type: file`, and `type: stdout` in TypeDB.
+* `logger`: Set up logging for modules in TypeDB, along with a log level and output targets (referencing outputs by name defined under the outputs section).
+* `debugger`: Set up TypeDB-specific debuggers. Right now, the only defined type is `type: reasoner`.
+
+## Configuration file options via command line arguments
+
+Use command line arguments to override any option in the configuration file.
+
+For example, the configuration file sets the server address as:
+
+<!-- test-ignore -->
+```yaml
+server:
+  address: 0.0.0.0:1729
+```
+
+If we want to use port 1730 instead of 1729, we can either update the configuration file or override it from the 
+command line using the following command:
+
+```bash
+typedb server --server.address 0.0.0.0:1730
+```
+
+Use the same approach to set a completely new section of the configuration that isn’t present in the file yet. For 
+example, to define a new logger subsection to print out all query plans, we could do the following to set the package 
+`com.vaticle.typedb.core.traversal` to output on a more verbose level:
+
+<!-- test-ignore -->
+```yaml
+typedb server  \
+  --server.address 0.0.0.0:1730  \
+  --log.logger.traversal.filter com.vaticle.typedb.core.traversal  \
+  --log.logger.traversal.level debug \
+  --log.logger.traversal.output "[ file, stdout ]"
+```
+
+## Cluster configuration
+
+Every server in a cluster has its own config file that contains a list of known servers in the cluster. A server in a 
+cluster will not accept connections from servers that are not in the list.
+
+<div class="note">
+[Note]
+Changes to the server configuration require a server restart to take effect.
+</div>
+
+To add or remove server to/from a cluster:
+
+1. Stop all TypeDB servers in the cluster.
+2. Update the configuration files of all (both new and old) TypeDB servers.
+3. Start all TypeDB servers which will be part of the new cluster.
```

Comment:
Is that even a feature? We basically recreate a cluster from scratch and data gets replicated.

Reply:
The raft paper has a section on how to do membership changes. 
I do not think just updating the config and letting the data replicate using the raft consensus is safe by itself, because I see a situation where an out of date server (from the initial set) may be elected the leader.

We should ideally verify this is safe (or figure out how to make it safe) before we publish this.

Reply:
Yes, we have not implemented the membership changes as advised by the Raft paper.

However, stopping the whole cluster, changing the config, and starting the whole cluster again should be safe, given that there's never any period of time where different machine has different consensus size.

What do you think @krishnangovindraj ?

Reply:
I'm unsure. The scenario that worries me is this:
Suppose we're resizing from 3 to 5. If , at the moment we bring the cluster down, two of the replicas are in sync and the third isn't, the latest committed entry will be replicated only on 2 of them.
When we bring them up with cluster-size set to 5, we are now in a state where the latest committed entry is not replicated to a majority - which I understand is a critical invariant.

I haven't checked this against the paper, but this might be sufficient to avoid the problem described:
1. Increment cluster sizes in hops of 2 at a time. 
2. Bring the new nodes up one at a time, and require that at least one entry is committed before we bring the next one up.


Of course, if all replicats are in sync when we bring the cluster down, it should be fine provided we're only increasing the size by 2

Reply:
@izmalk, let's exclude this section for now and wait until we implement proper cluster resizing. Proving the safety of this workaround has been taking some time, and it might be better just wait until we implement it properly!

Reply:
> @izmalk, let's exclude this section for now and wait until we implement proper cluster resizing

I would prefer to keep it and maybe provide a warning to do it in a safer way (like one new replica at a time or copy the stored data manually). Just because we don't have that many features for clusters. But this is the TypeDB section and not TypeDB Cloud, so we might as well comment it out for now. Thank you, everyone!